### PR TITLE
[EHL] Fix for Gbe device status for Fusa

### PR DESCRIPTION
**Pull request by koktong-ong**
TSB GBE PMECTRLSTATUS register returned 0x3 after
booting into OS and only occured in Fusa sku
This fix will check TSN PMECTRLSTATUS register and
update value to 0x0 if non-zero value was returned
Only applicable if TSN is enabled

Signed-off-by: Ong Kok Tong <kok.tong.ong@intel.com>

### CODE REVIEW

File: Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c [hunk 1]
```diff
@@ -49,6 +51,20 @@ typedef union {
   UINTN Data;
 } UINTN_STRUCT;
 
+VOID
```
Comment (myguitar) [inline Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c:54]: Add function description

File: Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c [hunk 2]
```diff
@@ -49,6 +51,20 @@ typedef union {
   UINTN Data;
 } UINTN_STRUCT;
 
+VOID
+CheckGbeStatus(
```
Comment (myguitar) [inline Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c:55]: A space `CheckGbeStatus( -> CheckGbeStatus (`

File: Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c [hunk 3]
```diff
@@ -49,6 +51,20 @@ typedef union {
   UINTN Data;
 } UINTN_STRUCT;
 
+VOID
+CheckGbeStatus(
+  UINTN GbeBase
+  )
+{
+  UINT16 pmectrlstatus;
```
Comment (myguitar) [inline Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c:59]: Can we use PmeCtrlStatus?

File: Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c [hunk 4]
```diff
@@ -49,6 +51,20 @@ typedef union {
...
 
+VOID
+CheckGbeStatus(
+  UINTN GbeBase
+  )
+{
+  UINT16 pmectrlstatus;
+  //Check device status
+  pmectrlstatus         = PciRead16 (GbeBase + PMECTRLSTATUS_OFFSET);
+  if (pmectrlstatus && POWERSTATE != 0){
```
Comment (myguitar) [inline Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c:62]: A space `0){ -> 0) {`

File: Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c [hunk 5]
```diff
@@ -49,6 +51,20 @@ typedef union {
...
+    PciWrite16((GbeBase + PMECTRLSTATUS_OFFSET), (pmectrlstatus & ~POWERSTATE));
```
Comment (myguitar) [inline Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c:63]: A space `PciWrite16(( -> PciWrite16 ((`

File: Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c [hunk 6]
```diff
@@ -126,14 +150,18 @@ IsPseGbe1Enabled (
...
   VendorId         = PciRead16 (PchPseGbeBase + PCI_VENDOR_ID_OFFSET);
 
   DEBUG ((DEBUG_INFO, "Pse Gbe 1 Device: Device Id: 0x%4X, Vendor Id: 0x%4X\n", DeviceId, VendorId));
   if ((DeviceId == 0xFFFF) || (VendorId == 0xFFFF)) {
     return FALSE;
   }
+  else{
+    //Check device status
+    CheckGbeStatus(PchPseGbeBase);
+  }
```
Comment (myguitar) [inline Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c:164]: Do we need this on Pse Gbe1? What about Gbe0?

**Review by myguitar — COMMENTED** [review]
Typo. in commit message? `TSB GBE PMECTRLSTATUS`

File: Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c [hunk 7]
```diff
@@ -49,6 +51,25 @@ typedef union {
...
+**/
+VOID
+CheckGbeStatus (
+  UINTN GbeBase
+  )
+{
+  UINT16 PmeCtrlStatus;
+  //Check device status
+  PmeCtrlStatus = PciRead16 (GbeBase + PMECTRLSTATUS_OFFSET);
+  if (PmeCtrlStatus && POWERSTATE != 0) {
```
Comment (myguitar) [inline Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c:67]: `&& -> &`
What is the exact condition here? To check both bits together? or any bit in POWERSTATE?
```
if (PmeCtrlStatus && POWERSTATE != 0) {
if ((PmeCtrlStatus & POWERSTATE) != POWERSTATE) {
```

File: Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c [hunk 8]
```diff
@@ -105,6 +130,10 @@ IsPseGbe0Enabled (
   if ((DeviceId == 0xFFFF) || (VendorId == 0xFFFF)) {
     return FALSE;
   }
+  else{
+     //Check device status
+     CheckGbeStatus(PchPseGbeBase);
```
Comment (myguitar) [inline Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c:135]: A space `CheckGbeStatus (PchPseGbeBase);`

File: Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c (hunk 7 above)
Comment (myguitar) [reply Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/UpdateAcpiGnvs.c]: Which condition is correct?

**Review by myguitar — APPROVED** [review]
Looks good to me

**Review by gdong1 — APPROVED** [review]

**Pull request merged**